File: test/test.EMGrank.c
#include "EMGrank.h"
#include "test_utils.h"
#include <stdlib.h>

int main(int argc, char** argv)
{
	int* dimensions = readArray_int("dimensions");
	int n = dimensions[0];
	int p = dimensions[1];
	int m = dimensions[2];
	int k = dimensions[3];
	free(dimensions);

	////////////
	// INPUTS //
	Real* Rho = readArray_real("Rho");
	Real* Pi = readArray_real("Pi");
	int mini = read_int("mini");
	int maxi = read_int("maxi");
	Real* X = readArray_real("X");
	Real* Y = readArray_real("Y");
	Real eps = read_real("eps");
	int* rank = readArray_int("rank");
	////////////

	/////////////
	// OUTPUTS //
	Real* phi = (Real*)malloc(p*m*k*sizeof(Real));
	Real* LLF = (Real*)malloc(1*sizeof(Real));
	/////////////

	//////////////////////////
	// Main call to EMGrank //
	EMGrank_core(Pi,Rho,mini,maxi,X,Y,eps,rank,
		phi,LLF,
		n,p,m,k);
	//////////////////////////

	free(Rho);
	free(Pi);
	free(X);
	free(Y);
	free(rank);

	// Compare to reference outputs
	Real* ref_phi = readArray_real("phi");
	compareArray_real("phi", phi, ref_phi, p*m*k);
	free(phi);
	free(ref_phi);

	// LLF
	Real* ref_LLF = readArray_real("LLF");
	compareArray_real("LLF", LLF, ref_LLF, 1);
	free(LLF);
	free(ref_LLF);

	return 0;
}

File: test/test_EMGrank.R
library(valse)
testFolder = "data/"

# NOTE: R typing is really terrible. as.double as.matrix ...and so on; don't remove.

#get inputs
npmk = as.matrix(read.table(paste(testFolder,"dimensions",sep="")))
n = npmk[1]; p=npmk[2]; m=npmk[3]; k=npmk[4]
Pi = as.double(as.matrix(read.table(paste(testFolder,"Pi",sep="")))[,])
Rho = array(as.double(as.matrix(read.table(paste(testFolder,"Rho",sep="")))), dim=c(m,m,k))
mini = as.integer(as.matrix(read.table(paste(testFolder,"mini",sep="")))[1])
maxi = as.integer(as.matrix(read.table(paste(testFolder,"maxi",sep="")))[1])
X = matrix(as.double(as.matrix(read.table(paste(testFolder,"X",sep="")))), n,p)
Y = matrix(as.double(as.matrix(read.table(paste(testFolder,"Y",sep="")))), n,m)
eps = as.double(as.matrix(read.table(paste(testFolder,"eps",sep="")))[1])
rank = as.double(as.matrix(read.table(paste(testFolder,"rank",sep="")))[,])

#get outputs
phi = array(as.double(as.matrix(read.table(paste(testFolder,"phi",sep="")))), dim=c(p,m,k))
LLF = as.double(as.matrix(read.table(paste(testFolder,"LLF",sep="")))[1])

res = valse::EMGrank(Pi,Rho,mini,maxi,X,Y,eps,rank,fast=TRUE)

#compare outputs
nd=7 #number of digits
print( all(round(phi,nd) == round(res$phi,nd)) )
print( all(round(LLF,nd) == round(res$LLF,nd)) )
